### PR TITLE
Refactor listing.go package for memory optimizations

### PR DESCRIPTION
**Pull request by amirmalka**
This pull request modifies the `AllLists` struct and its associated methods to improve the memory utilization of Kubescape.

A detailed breakdown of the changes:

1. The original `AllLists` struct used separate slices for different statuses (`passed`, `failed`, `skipped`, and `other`). The new version replaces these with a single map `itemToStatus`, where the key is the item (control ID / resource ID), and the value is the status. The change from slices to a map allows for efficient constant-time lookups and modifications for individual elements based on their keys.

2. Two new functions `GetAllListsFromPool()` and `PutAllListsToPool()` are added, which manage an `AllLists` object pool. This pool reuses the `AllLists` objects, reducing the overhead of memory allocations and deallocations.

3. The `Initialize()` function preallocates the map with a given size for optimization purposes. The `Clear()` function resets the map and all counters - needed when getting `AllLists` from the object pool, as it might be dirty.

4. The `Append()` method is updated to work with the new `itemToStatus` map. It now checks if the item already exists with a different status. If so, it compares the significance of the old and new statuses. The item's status gets updated if the new status is more significant, and the corresponding status counters are adjusted accordingly.  With the new logic, we no longer need the `ToUniqueResources` and `ToUniqueControls` methods as they were used to remove duplicates and ensure the uniqueness of items within the `AllLists` structure.

5. The `Update()` method merges another `AllLists` into the current one by simply calling the `Append()` on each item - this operation ensures that statuses are updated correctly based on their significance if items are present in both lists.

6. The `GetItems()` method retrieves all items with a particular status. It replaces the old `Passed()`, `Failed()`, `Skipped()`, and `Other()` methods.

7. Separate counters (`passed`, `failed`, `skipped`, and `other`) are kept for each status, providing constant-time access to the counts.

8. Removed the Iterator interface and the AllListsIterator struct which were used to sequentially access elements in the different status lists (failed, passed, skipped, other) - in the updated code, we simply return the `itemToStatus` map so it can be iterated without a special iterator. 

------

Memory profiles taken on Kubescape scan, before and after the proposed optimizations
*K8S cluster size ~15,000 resources*

**BEFORE**

Screenshot 2023-05-23 at 10 15 08

**AFTER**

Screenshot 2023-05-23 at 10 14 58

-------

Allocs profile focused on the `updateResults` method

**BEFORE**

Screenshot 2023-05-23 at 10 17 14

**AFTER**

Screenshot 2023-05-23 at 10 17 20

### CODE REVIEW

**Review by dwertent — APPROVED** [review]
Well done!!!

File: reporthandling/helpers/v1/listing.go [hunk 1]
```diff
@@ -1,145 +1,154 @@
...
 
+var allListsPool = &sync.Pool{
+	New: func() interface{} {
+		return &AllLists{}
+	},
+}
+
+// GetAllListsFromPool get the AllLists object from the pool
+func GetAllListsFromPool() *AllLists {
+	l := allListsPool.Get().(*AllLists)
```
Comment (RinaO1234) [inline reporthandling/helpers/v1/listing.go:18]: can we have ,ok check to verify conversion is ok?

File: reporthandling/helpers/v1/listing.go [hunk 2]
```diff
@@ -1,145 +1,154 @@
...
+				case apis.StatusFailed:
+					all.failed++
+				case apis.StatusSkipped:
+					all.skipped++
+				default:
+					all.other++
+				}
+
+				// update the old status
+				switch oldStatus {
```
Comment (RinaO1234) [inline reporthandling/helpers/v1/listing.go:107]: this logic fits a new function to get the status and int to append ( the count with some int (+/-))

File: reporthandling/helpers/v1/listing.go (hunk 1 above)
Comment (amirmalka) [reply reporthandling/helpers/v1/listing.go:18]: https://github.com/kubescape/opa-utils/pull/108

File: reporthandling/helpers/v1/listing.go (hunk 2 above)
Comment (amirmalka) [reply reporthandling/helpers/v1/listing.go:107]: Fixed in https://github.com/kubescape/opa-utils/pull/108

**Pull request merged**